### ISSUE TITLE
High latency on audio-only pipeline

### ISSUE THREAD

**Issue by ikonst** [EricssonResearch/openwebrtc#155]
This issue is to track latency issues. I'm focusing on my iOS test app "TestSendReceive", which basically sends PCMA to localhost (with DTLS off).

I'm hearing echo in ~ 1 second, which is too high.

**Comment by superdump** [comment]
I think there are avaudiosession configuration categories that may affect capture and rendering latency. As I recall if you check the gst debug output you should see what is being used for latency settings in the sink and source. There should not be significant latency in any of the pipelines but if there is, it's a bug.

**Comment by ikonst** [comment]
I'm setting up AVAudioSession with:
````objc
    [theSession setPreferredSampleRate:audio_sample_rate  error:nil];
    [theSession setMode:AVAudioSessionModeVideoChat error:nil];
    [theSession setCategory:AVAudioSessionCategoryPlayAndRecord error:nil];
````

BTW, `AVAudioSessionModeVideoChat` affects choice of microphone and equalizer, not latency.

`mySession.IOBufferDuration + mySession.inputLatency` add ~0.028 seconds of latency.

The research continues.

**Comment by ikonst** [comment]
`gst_osx_audio_src_io_proc` is called with `inNumberFrames = 128` which doesn't sound too high either.

**Comment by ikonst** [comment]
On a "self-view" audio iOS app (source → renderer, without transport) I get unacceptable latency of ~ 0.5s (or maybe less - I measured by ear - but it's noticeable, and it shouldn't be).

**Comment by ikonst** [comment]
@superdump , @sdroege - you had some observations today about the InterAudioSink/Src adding to the latency

**Comment by superdump** [comment]
Yes. interaudiosrc currently has s default latency-time value of 100ms. Latency-time cannot be less than period-time and period-time should probably be an audio frame, which is often 20ms but depends on codec if a codec is involved.

I feel like for a low latency pipelines the interaudio* elements should probably act like a queue in order to not add any latency. @sdroege suggested this could be implemented as a mode for interaudio* pretty quickly. I have been looking over some pipeline graphs too and will discuss them a bit tomorrow as I have been experiencing quite high latency in some situations. I will prioritize looking into this and resolving it.

**Comment by sdroege** [comment]
I'll look into implementing a "waiting mode" in interaudiosrc this week, which does not solve any problem by itself but will reduce the number of knobs you have for controlling the latency. It's slightly more complicated than I thought yesterday though, needs some more thinking

**Comment by superdump** [comment]
@sdroege ping?

**Comment by ford-prefect** [comment]
I'm looking at this in parallel with the OS X patches.

**Comment by sdroege** [comment]
I have some ideas how we can improve latency handling (and at the same time make zerocopy between camera/encoder/sink or camera/decoder/sink work better). Will write that down tomorrow for further discussion.
